### PR TITLE
Make some source cleanup

### PR DESCRIPTION
**Pull request by pablogsal**
Update the source files to use C++ idioms, remove unused headers, remove
 shadowing variables, remove empty declarations, make member functions
 static when possible and fix small bugs due to calling virtual functions
in destructors.

### CODE REVIEW

File: src/memray/_memray/hooks.cpp [hunk 1]
```diff
@@ -279,7 +276,7 @@ dlopen(const char* filename, int flag) noexcept
     void* ret = hooks::dlopen(filename, flag);
     if (ret) {
         tracking_api::Tracker::invalidate_module_cache();
-        if (filename && 0 != strstr(filename, "/_greenlet.")) {
+        if (filename && nullptr != strstr(filename, "/_greenlet.")) {
```
Comment (lkollar) [inline src/memray/_memray/hooks.cpp:279]: Is this change necessary? My understanding is that `strstr` returns a null pointer if there is no match.

Comment (pablogsal) [reply src/memray/_memray/hooks.cpp:279]: Yeah, that's what I am changing :)

**Review by lkollar — APPROVED** [review]
